### PR TITLE
Mmillet/spi 515 add all secrets option to ggshield secret scans

### PR DESCRIPTION
**Pull request by gg-mmill**
- Add the `--all-secrets` option, such that ignore options are "ignored": all secrets are returned, with `is_excluded` / `exclude_reason` (sic) set appropriately in the policy breaks.
- Adds a message when excluded secrets are found, to advertise the feature
- Serialize secret options in a header, for option usage monitoring

## What has been done

- add `--all-secret` option
- update output handlers
    - json output is always extended
    - text output
         -   when option is not passed, add line below file info showing number of hidden secrets
         -  when option is passed, show all secrets, and add `Exclude reason` when appropriate

## Validation

Run ggshield scans on a file containing secrets with ignored / resolved issues, with and without `--all-secrets` option

## PR check list

- [ ] As much as possible, the changes include tests (unit and/or functional)
- [ ] If the changes affect the end user (new feature, behavior change, bug fix) then the PR has a changelog entry (see doc/dev/getting-started.md). If the changes do not affect the end user, then the `skip-changelog` label has been added to the PR.

### CODE REVIEW

File: ggshield/core/config/user_config.py [hunk 1]
```diff
@@ -61,6 +63,22 @@ def add_ignored_match(self, secret: IgnoredMatch) -> None:
...
         self.ignored_matches.append(secret)
 
+    def dump_for_monitoring(self) -> str:
+        return json.dumps(
+            {
+                "show_secrets": self.show_secrets,
+                "ignored_detectors_count": len(self.ignored_detectors),
+                "ignored_matches_count": len(self.ignored_matches),
+                "ignored_paths_count": len(self.ignored_paths),
+                "ignore_known_secrets_secrets": self.show_secrets,
```
Comment (agateau-gg) [inline ggshield/core/config/user_config.py:73]: I think this should be:
```suggestion
                "show_secrets": self.show_secrets,
```

**Review by agateau-gg — CHANGES_REQUESTED** [review]
Minor remarks, but this looks good, and the addition of factories is great!

File: ggshield/verticals/secret/secret_scan_collection.py [hunk 2]
```diff
@@ -90,21 +89,41 @@ def censor(self) -> None:
     def has_policy_breaks(self) -> bool:
         return len(self.policy_breaks) > 0
 
-    def apply_ignore_function(
-        self, reason: IgnoreReason, ignore_function: Callable[[PolicyBreak], bool]
+    @classmethod
+    def from_scan_result(
+        cls, file: Scannable, scan_result: ScanResult, secret_config: SecretConfig
     ):
```
[the remark this answers is not in the document]
Comment (gg-mmill) [reply ggshield/verticals/secret/secret_scan_collection.py:95]: I don't fully understand why, but pyright is unhappy (with `import Self from typing_extensions`):
` - error: Expression of type "Result" is incompatible with return type "Self@Result";`

Using `Result` instead.

File: ggshield/verticals/secret/secret_scan_collection.py [hunk 3]
```diff
@@ -1,36 +1,56 @@
...
+
 
+def compute_ignore_reason(
+    policy_break: PolicyBreak, secret_config: SecretConfig
+) -> Union[str, None]:
+    """Computes the possible ignore reason associated with a PolicyBreak"""
+    ignore_reason = None
+    if policy_break.diff_kind in {DiffKind.DELETION, DiffKind.CONTEXT}:
+        ignore_reason = IgnoreReason.NOT_INTRODUCED
+    if policy_break.is_excluded:
```
[the remark this answers is not in the document]
Comment (gg-mmill) [reply ggshield/verticals/secret/secret_scan_collection.py]: Was not on purpose, thanks for noticing !

File: tests/unit/verticals/secret/test_secret_scan_collection.py [hunk 4]
```diff
@@ -19,3 +40,147 @@ def test_results_from_exception():
...
+    @pytest.mark.parametrize("ignore_known", (True, False))
+    def test_known_secret(self, ignore_known):
+        """
+        GIVEN a known policy break
+        WHEN computing the ignore reason
+        THEN it's not None iff ignore_secret is enabled in config
+        """
+        policy_break = PolicyBreakFactory(known_secret=True)
+        config = SecretConfig(ignore_known_secrets=ignore_known)
+        assert (compute_ignore_reason(policy_break, config) is not None) is ignore_known
```
[the remark this answers is not in the document]
Comment (gg-mmill) [reply tests/unit/verticals/secret/test_secret_scan_collection.py]: Had to go with `assert bool(compute_ignore_reason(policy_break, config)) is ignore_known`

**Pull request merged**